apps/notifications/src/panel/Notifications.jsx

```diff
@@ -48,40 +48,42 @@ export class Notifications extends PureComponent {
 		receiveMessage: noop,
 	};
 
+	defaultHandlers = {
+		APP_REFRESH_NOTES: [
+			( _store, action ) => {
+				if ( ! client ) {
+					return;
+				}
+
+				if ( 'boolean' === typeof action.isVisible ) {
+					debug( 'APP_REFRESH_NOTES', {
+						isShowing: this.props.isShowing,
+						isVisible: action.isVisible,
+					} );
+					// Use this.props instead of destructuring isShowing, so that this uses
+					// the value on props at any given time and not only the value that was
+					// present on initial mount.
+					client.setVisibility.call( client, {
+						isShowing: this.props.isShowing,
+						isVisible: action.isVisible,
+					} );
+				}
+
+				client.refreshNotes.call( client, action.isVisible );
+			},
+		],
+	};
+
 	constructor( props ) {
 		super( props );
 
 		const { customEnhancer, customMiddleware, isShowing, isVisible, receiveMessage, wpcom } =
 			this.props;
 
-		initStore( {
-			customEnhancer,
-			customMiddleware: mergeHandlers( customMiddleware, {
-				APP_REFRESH_NOTES: [
-					( _store, action ) => {
-						if ( ! client ) {
-							return;
-						}
-
-						if ( 'boolean' === typeof action.isVisible ) {
-							debug( 'APP_REFRESH_NOTES', {
-								isShowing: this.props.isShowing,
-								isVisible: action.isVisible,
-							} );
-							// Use this.props instead of destructuring isShowing, so that this uses
-							// the value on props at any given time and not only the value that was
-							// present on initial mount.
-							client.setVisibility.call( client, {
-								isShowing: this.props.isShowing,
-								isVisible: action.isVisible,
-							} );
-						}
-
-						client.refreshNotes.call( client, action.isVisible );
-					},
-				],
-			} ),
-		} );
+		initStore( { customEnhancer } );
```

Comment:
This `initStore` call has any effect only in the `standalone` app, where the `customEnhancer` will assign the store to a local `store` variable. But the `initStore` function no longer assigns the store to the local `store` variable in the `state` module. It's now pure. But there are many modules, (`FilterBarController`, even the `Notifications` module I'm commenting on), that `import { store } from './state'`. They use the "wrong" store.

Reply:
One specific way how the `standalone` app will be failing now is when receiving messages from the parent window. The message handlers (`refresh`, `reset`) will dispatch actions to the local store, but the rest of the app is using the store in the `state` module.

A good solution would be to stop setting the local store, and just import `store` from `../panel/state`.

The end result is that all `Notifications` panels use the same instance of Redux store, but only add and remove their own listeners when mounting and unmounting.

Reply:
Is there any use case where we want two different `<Notifications>` instances to be rendered simultaneously? I think that's never supported and will break when we do so. That's why I added an additional check in this PR to avoid rendering the global notifications panel when visiting the `/read/notifications`.

> A good solution would be to stop setting the local store, and just import `store` from `../panel/state`.

I think a better solution might be the other way around. Since `connect` and `useSelector` always use the local store from the `<Provider>`. However, I think either way works if we ensure that there's only going to be one store at a time.

Reply:
> Is there any use case where we want two different <Notifications> instances to be rendered simultaneously?

There was the `/read/notifications` page, but that's now fixed. But what remains is that two different `<Notifications>` instances, not simultaneously but one after another, will use the same Redux store. And they need to work together well to prevent bugs like the `n` key not working.

> I think a better solution might be the other way around. Since `connect` and `useSelector` always use the local store from the `<Provider>`.

Yes, I agree this is the final state to which we should eventually refactor. Two things that currently prevent this:
- in the `standalone` app, the `store` is used _above_ the provider, via a global variable
- this global variable is imported at ~5 places (`import { store } from '../state'`). All these usages would need to be refactored to `react-redux` APIs that read the `store` from context

If you want to do this in this PR, I'm all for it. Otherwise the `standalone` app needs to be fixed some other way.

Reply:
> But what remains is that two different `<Notifications>` instances, not simultaneously but one after another, will use the same Redux store. And they need to work together well to prevent bugs like the `n` key not working.

Could you elaborate how that won't work with reproduction steps? I think that's what `removeListeners` fixed when the component unmounts.

> If you want to do this in this PR, I'm all for it.

I want to avoid that in this PR and focus on fixing the issue first. 😅

Reply:
> Could you elaborate how that won't work with reproduction steps?

`removeListeners` fixes the popup inside Calypso and its conflict with `/read/notifications`. But I believe that it will be the `standalone` app that will break, namely when receiving messages from the parent window. I can prepare specific reproduction steps, I'm curious about that myself 🙂

Reply:
> I can prepare specific reproduction steps

The code is so convoluted that I had to give up reproducing something 😅 Two tiny things that are maybe broken are:
- on closing the notification panel, selected notification should be reset: if you reopen, nothing should be selected
- when the panel closes or the browser window becomes invisible (`document.visibilityState`), the notifications shouldn't be refreshed in background

But both patched and unpatched versions work the same way, maybe these small features are already broken? I didn't find anything that should block this PR.